src/mstr.h

```diff
@@ -0,0 +1,223 @@
+ * that aggregates several keyspace related metadata into one compact, singly
+ * allocated, mstring.
+ *
+ *      typedef enum HkeyMetaFlags {
+ *          HKEY_META_VAL_REF_COUNT    = 0,  // refcount
+ *          HKEY_META_VAL_REF          = 1,  // Val referenced
+ *          HKEY_META_EXPIRE           = 2,  // TTL and more
+ *          HKEY_META_TYPE_ENC_LRU     = 3,  // TYPE + LRU + ENC
+ *          HKEY_META_DICT_ENT_NEXT    = 4,  // Next dict entry
+ *          // Following two must be together and in this order
+ *          HKEY_META_VAL_EMBED8       = 5,  // Val embedded, max 7 bytes
+ *          HKEY_META_VAL_EMBED16      = 6,  // Val embedded, max 15 bytes (23 with EMBED8)
+ *      } HkeyMetaFlags;
+ *
+ *      mstrKind hkeyKind = {
+ *          .name = "hkey",
+ *          .metaSize[HKEY_META_VAL_REF_COUNT] = 4,
+ *          .metaSize[HKEY_META_VAL_REF]       = 8,
+ *          .metaSize[HKEY_META_EXPIRE]        = sizeof(ExpireMeta),
+ *          .metaSize[HKEY_META_TYPE_ENC_LRU]  = 8,
+ *          .metaSize[HKEY_META_DICT_ENT_NEXT] = 8,
+ *          .metaSize[HKEY_META_VAL_EMBED8]    = 8,
+ *          .metaSize[HKEY_META_VAL_EMBED16]   = 16,
+ *      };
+ *
+ * MSTR-ALIGNMENT
+ * --------------
+ * There are two types of alignments to take into consideration:
+ * 1. Alignment of the metadata.
+ * 2. Alignment of returned mstr pointer
+ *
+ * 1) As the metadatas layout are reversed to their enumeration, it is recommended
+ *    to put metadata with "better" alignment first in memory layout (enumerated
+ *    last) and the worst, or those that simply don't require any alignment will be
+ *    last in memory layout (enumerated first). This is similar the to the applied
+ *    consideration when defining new struct in C. Note also that each metadata
+ *    might either be attached to mstr or not which complicates the design phase
+ *    of a new mstrKind a little.
+ *
+ *    In the example above, HKEY_META_VAL_REF_COUNT, with worst alignment of 4
+ *    bytes, is enumerated first, and therefore, will be last in memory layout.
+ *
+ * 2) Few optimizations in Redis rely on the fact that sds address is always an odd
+ *    pointer. We can achieve the same with a little effort. It was already taken
+ *    care that all headers of type mstrhdrX has odd size. With that in mind, if
+ *    a new kind of mstr is required to be limited to odd addresses, then we must
+ *    make sure that sizes of all related metadatas that are defined in mstrKind
+ *    are even in size.
+ */
+
+#ifndef __MSTR_H
+#define __MSTR_H
+
+#include <sys/types.h>
+#include <stdarg.h>
+#include <stdint.h>
+
+/* Selective copy of ifndef from server.h instead of including it */
+#ifndef static_assert
+#define static_assert(expr, lit) extern char __static_assert_failure[(expr) ? 1:-1]
+#endif
+
+#define MSTR_TYPE_5         0
+#define MSTR_TYPE_8         1
+#define MSTR_TYPE_16        2
+#define MSTR_TYPE_64        3
+#define MSTR_TYPE_MASK      3
+#define MSTR_TYPE_BITS      2
```

Comment:
This means that we won't be able to add new type in the future.

Reply:
At `sds` it is just the same. Unlike `sds` we can use here the flags and attached metadata to indicate properly any future requirements.